Add LocalWaker and ContextBuilder types to core, and LocalWake trait to alloc.

tvallotton opened the pull request:
Implementation for  #118959.

tvallotton committed:
feat: add LocalWaker type, ContextBuilder type, and LocalWake trait.

tvallotton committed:
perf: move null check from local_wake() to build()

tvallotton committed:
refactor: remove  in favor of  and  to make the API infallible.

tvallotton committed:
feat: add try_waker and From<&mut Context> for ContextBuilder to allow the extention of contexts by futures

tvallotton committed:
fix: make LocalWake available in targets that don't support atomics by removing a #[cfg(target_has_atomic = ptr)]

tvallotton committed:
chore: add tracking issue number to local waker feature

tvallotton committed:
feat: impl AsRef<LocalWaker> for Waker.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7217153045/job/19664501265) [(plain)](https://github.com/rust-lang/rust/commit/3343703c8de9323e9b7735180fb7d00d987b6939/checks/19664501265/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=tvallotton
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_df7b076d-f404-44a9-9b7c-8ce301336e72
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=local_waker
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_df7b076d-f404-44a9-9b7c-8ce301336e72
GITHUB_REF=refs/pull/118960/merge
GITHUB_REF_NAME=118960/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=3ffc07e341c81fae01dcaa254339a2cb43fa16cd
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_df7b076d-f404-44a9-9b7c-8ce301336e72
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_df7b076d-f404-44a9-9b7c-8ce301336e72
GITHUB_TRIGGERING_ACTOR=tvallotton
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118960/merge
GITHUB_WORKFLOW_SHA=3ffc07e341c81fae01dcaa254339a2cb43fa16cd
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container f4e1de0e3b22
 ---> a09069b38488
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> d4bdb0098040
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-wovww8tv/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.25.0-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 19.9 MB/s eta 0:00:00
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 45.8 MB/s eta 0:00:00
Collecting platformdirs<5,>=3.9.1
  Downloading platformdirs-4.1.0-py3-none-any.whl (17 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.8 filelock-3.13.1 platformdirs-4.1.0 virtualenv-20.25.0
Removing intermediate container a691700ba850
 ---> 446bebd22464
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> cd1b0346e3f7
 ---> cd1b0346e3f7
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> be674bf0b785
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container c759c07f97db
 ---> 607f7db3cb8e
Successfully built 607f7db3cb8e
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:607f7db3cb8e747bdd6d1a83a12812200ea886797d82fc6e9bddfa2e861bfde5
Uploading finished image sha256:607f7db3cb8e747bdd6d1a83a12812200ea886797d82fc6e9bddfa2e861bfde5 to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
607f7db3cb8e   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
cd1b0346e3f7   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
446bebd22464   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
d4bdb0098040   10 seconds ago   /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
a09069b38488   11 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      13 days ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      13 days ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      13 days ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7fd93bd39b90>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:607f7db3cb8e747bdd6d1a83a12812200ea886797d82fc6e9bddfa2e861bfde5
---
DirectMap4k:      206784 kB
DirectMap2M:     8181760 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/03515c6a2278aba8d64def5c522fb146e721368f/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-03515c6a2278aba8d64def5c522fb146e721368f-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 25.66s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/core/src/task/wake.rs at line 269:
 /// #![feature(noop_waker)]
 /// use std::task::{ContextBuilder, LocalWaker, Waker, Poll};
 /// use std::future::Future;
-/// 
+///
 /// let local_waker = LocalWaker::noop();
 /// let waker = Waker::noop();
 ///
##[error]Diff in /checkout/library/core/src/task/wake.rs at line 347:
 /// use std::task::{Waker, ContextBuilder};
 /// use std::future::{poll_fn, Future};
 /// use std::pin::pin;
-/// 
+///
 /// async fn with_waker<F>(f: F, waker: &Waker) -> F::Output
 /// where
 ///     F: Future
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/str/error.rs" "/checkout/library/core/src/ascii.rs" "/checkout/library/core/src/task/ready.rs" "/checkout/library/core/src/task/wake.rs" "/checkout/library/core/src/task/poll.rs" "/checkout/library/core/src/task/mod.rs" "/checkout/library/core/src/num/mod.rs" "/checkout/library/core/src/str/traits.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri Dec 15 02:26:38 UTC 2023
  network time: Fri, 15 Dec 2023 02:26:39 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

tvallotton committed:
chore: fix ci failures

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7217288779/job/19664856406) [(plain)](https://github.com/rust-lang/rust/commit/6d66430b7f0ff5cda84892dde65058b19f530813/checks/19664856406/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=tvallotton
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_6134b74b-9c44-443c-8c7e-09991b31c219
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=local_waker
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_6134b74b-9c44-443c-8c7e-09991b31c219
GITHUB_REF=refs/pull/118960/merge
GITHUB_REF_NAME=118960/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=24ca082d655bcc2ded0ab710114bc757a50ccb43
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_6134b74b-9c44-443c-8c7e-09991b31c219
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_6134b74b-9c44-443c-8c7e-09991b31c219
GITHUB_TRIGGERING_ACTOR=tvallotton
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118960/merge
GITHUB_WORKFLOW_SHA=24ca082d655bcc2ded0ab710114bc757a50ccb43
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container 8543f0e88c3a
 ---> 35d43637ae64
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> f3d7ce093853
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-sqttxd_m/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
  Downloading virtualenv-20.25.0-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 54.0 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Collecting platformdirs<5,>=3.9.1
  Downloading platformdirs-4.1.0-py3-none-any.whl (17 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 108.1 MB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.8 filelock-3.13.1 platformdirs-4.1.0 virtualenv-20.25.0
Removing intermediate container a48d1c075165
 ---> 842bfd5d039f
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 59afade07fc0
 ---> 59afade07fc0
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 7d73cf974344
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container def45cacad21
 ---> 062f11a690da
Successfully built 062f11a690da
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:062f11a690daf2e65c4b7acc1933754db5fc9f037292f94363d3b9c61542e55f
Uploading finished image sha256:062f11a690daf2e65c4b7acc1933754db5fc9f037292f94363d3b9c61542e55f to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
062f11a690da   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
59afade07fc0   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
842bfd5d039f   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
f3d7ce093853   9 seconds ago    /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
35d43637ae64   10 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      13 days ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      13 days ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      13 days ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f3748985e50>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:062f11a690daf2e65c4b7acc1933754db5fc9f037292f94363d3b9c61542e55f
---
DirectMap4k:      182208 kB
DirectMap2M:     6109184 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/03515c6a2278aba8d64def5c522fb146e721368f/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-03515c6a2278aba8d64def5c522fb146e721368f-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/core/src/task/wake.rs:713: `issue` "87021" mismatches the previous `issue` of "96992"
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (23.3.1)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 41.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 41.4 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 148.0 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:01:03
  local time: Fri Dec 15 02:46:14 UTC 2023
  network time: Fri, 15 Dec 2023 02:46:14 GMT
```

</details>

cynecx commented on library/core/src/task/wake.rs:
```diff
@@ -418,3 +571,216 @@ impl fmt::Debug for Waker {
+#[unstable(feature = "local_waker", issue = "118959")]
+#[repr(transparent)]
+pub struct LocalWaker {
+    waker: RawWaker,
```
I get that this implicitly implies `LocalWaker: !Send, !Sync`. But does it make sense to spell out this stability guarantee with:

```rust
impl !Send for LocalWaker {}
impl !Sync for LocalWaker {}
```

tvallotton replied:
I guess that would make sense, I forgot `std` could have negative implementations.

tvallotton replied:
It was added on https://github.com/rust-lang/rust/pull/118960/commits/017fe2f03170994524edf658d43ad7e44b2f8db2.

tvallotton commented:
This CI failure is strange: 
> Error: tidy error: /checkout/library/core/src/task/wake.rs:713: `issue` "87021" mismatches the previous `issue` of "96992"

It seems the error message is correct about the fact that we shouldn't be linking to #87021 and we should link instead to #96992. But it makes me question why does it only fail on my branch, I didn't change this issue number after all.

tvallotton committed:
fix: change issue number of waker_getters from #87021 to #96992.

tvallotton committed:
chore: add and !Sync impls for LocalWaker as a stability guarantee.

tvallotton committed:
doc: update thread safety explanation for RawWakerVTable and RawWaker.

tvallotton commented:
Just for the record, it might be wise to check if `ContextBuilder<'a>` should be future proofed to be invariant, since I noticed that the `with_waker` example seems to break when changing the variance. Naturally, that makes the implementation of `From<&mut Context>` practically useless.

Mark-Simulacrum commented on library/alloc/src/task.rs:
```diff
@@ -152,3 +159,167 @@ fn raw_waker<W: Wake + Send + Sync + 'static>(waker: Arc<W>) -> RawWaker {
+/// to hold data that does not implement `Send` and `Sync`. Additionally, it saves calls
+/// to `Arc::clone`, which requires atomic synchronization.
+///
+
```
```suggestion
///
```

Mark-Simulacrum commented on library/alloc/src/task.rs:
```diff
@@ -152,3 +159,167 @@ fn raw_waker<W: Wake + Send + Sync + 'static>(waker: Arc<W>) -> RawWaker {
+/// is used to push new tasks onto the run queue, while the block on function will remove them
+/// and poll them. When a task is woken, it will put itself back on the run queue to be polled by the executor.
+///
+/// **Note:** A real world example would interlieve poll calls with calls to an io reactor to wait for events instead
```
```suggestion
/// **Note:** A real world example would interleave poll calls with calls to an io reactor to wait for events instead
```

Mark-Simulacrum commented on library/alloc/src/task.rs:
```diff
@@ -152,3 +159,167 @@ fn raw_waker<W: Wake + Send + Sync + 'static>(waker: Arc<W>) -> RawWaker {
+/// {
+///     spawn(future);
+///     loop {
+///         let Some(task) = RUN_QUEUE.with_borrow_mut(|queue|queue.pop_front()) else {
```
```suggestion
///         let Some(task) = RUN_QUEUE.with_borrow_mut(|queue| queue.pop_front()) else {
```

Mark-Simulacrum commented on library/core/src/task/wake.rs:
```diff
@@ -33,9 +35,18 @@ impl RawWaker {
+    /// thread safe type such as an `[Arc]<T: Send + Sync>`
+    /// when used to construct a [`Waker`]. This restriction is lifted when
+    /// constructing a [`LocalWaker`], which allows using types that do not implement
+    /// <code>[Send] + [Sync]</code> like `[Rc]<T: !Send + !Sync>`.
```
`Rc<T>` is never Send/Sync, regardless of the T within, so this description feels a bit off to me. Maybe just drop those bounds here?

Mark-Simulacrum commented on library/core/src/task/wake.rs:
```diff
@@ -418,3 +594,222 @@ impl fmt::Debug for Waker {
+/// [`local_waker`]: core::task::Context::local_waker
+
+#[unstable(feature = "local_waker", issue = "118959")]
+#[repr(transparent)]
```
```suggestion
#[cfg_attr(not(doc), repr(transparent))] // work around https://github.com/rust-lang/rust/issues/66401
```

Probably makes sense to match Waker.

Mark-Simulacrum commented on library/core/src/task/wake.rs:
```diff
@@ -215,6 +283,125 @@ impl fmt::Debug for Context<'_> {
+    #[rustc_const_unstable(feature = "const_waker", issue = "102012")]
+    #[unstable(feature = "local_waker", issue = "118959")]
+    pub const fn waker(self, waker: &'a Waker) -> Self {
+        Self { waker: Some(waker), ..self }
```
It feels strange that `from_waker` implicitly sets the local_waker, but `waker(waker)` does not. Can we at least highlight that difference in documentation? While this is unstable surface area it doesn't matter too much, I suppose.

tvallotton replied:
We could change it to work like this: 
```rust
 pub fn waker(self, waker: &'a Waker) -> Self {
         let Some(waker) = self.waker else {
            return Self { waker: Some(waker), ..self };
        };

        let override_waker = waker.as_ref().will_wake(self.local_waker);

        if !override_waker {
            return Self { waker: Some(waker), ..self };
        }
        // SAFETY: LocalWaker is just Waker without thread safety
        let local_waker = unsafe { transmute(waker) };
        return Self { waker: Some(waker), local_waker, ..self };
    }
```
This way, the local waker will only be overridden if it hasn't been specialized yet. But then the API would seize to be `const`, unless we are willing to compare them like this: 
```rust
let override_waker = unsafe {
      transmute::<_, [usize; 2]>(waker) != transmute::<_, [usize; 2]>(self.local_waker)
};
```
What do you think it's best?

tvallotton replied:
I do think it is unlikely users will set the waker twice, so I wouldn't really mind staying with the original implementation.

Mark-Simulacrum commented on library/core/src/task/wake.rs:
```diff
@@ -215,6 +283,125 @@ impl fmt::Debug for Context<'_> {
+    }
+
+    /// This field is used to set the value of the waker on `Context`.
+
```
```suggestion
```

Mark-Simulacrum commented on library/core/src/task/wake.rs:
```diff
@@ -215,6 +283,125 @@ impl fmt::Debug for Context<'_> {
+impl<'a> From<&mut Context<'a>> for ContextBuilder<'a> {
+    #[inline]
+    fn from(value: &mut Context<'a>) -> Self {
+        let Context { waker, local_waker, .. } = *value;
```
`..` doesn't feel appropriate here, I'd rather see the fields explicitly written out so any additional fields get caught at compile time as needing changes here.

tvallotton replied:
Done

Mark-Simulacrum commented on library/core/src/task/wake.rs:
```diff
@@ -418,3 +594,222 @@ impl fmt::Debug for Waker {
 }
+
+/// A `LocalWaker` is analogous to a [`Waker`], but it does not implement [`Send`] or [`Sync`].
+/// This handle encapsulates a [`RawWaker`] instance, which defines the
```
```suggestion
///
/// This handle encapsulates a [`RawWaker`] instance, which defines the
```

Mark-Simulacrum commented on library/core/src/task/wake.rs:
```diff
@@ -418,3 +594,222 @@ impl fmt::Debug for Waker {
+/// [`Future::poll()`]: core::future::Future::poll
+/// [`Poll::Pending`]: core::task::Poll::Pending
+/// [`local_waker`]: core::task::Context::local_waker
+
```
```suggestion
```

Mark-Simulacrum commented on library/core/src/task/wake.rs:
```diff
@@ -418,3 +594,222 @@ impl fmt::Debug for Waker {
+    /// calling `waker.clone().wake()`.
+    #[inline]
+    #[stable(feature = "futures_api", since = "1.36.0")]
+    pub fn wake_by_ref(&self) {
```
Can we keep the order of methods consistent with Waker? i.e., moving will-wake and wake-by-ref up?

Mark-Simulacrum commented on library/core/src/task/wake.rs:
```diff
@@ -195,16 +236,43 @@ impl<'a> Context<'a> {
+    ///
+    /// Note that if the waker does not need to be sent across threads, it
+    /// is preferable to call `local_waker`, which is more portable and
+    /// potentially more efficient.
```
I'm not sure we should say this while the API remains unstable. I'm not sure how much precedent we have, but I feel like we generally try to avoid docs talking too much about unstable area. This seems like a particularly clear case where *right now* it's quite unlikely that any of the surface area added in this PR is what you should use "by default".

Mark-Simulacrum reviewed (commented):
Do we have signoff from T-libs-api on the new unstable surface area? The ACP issue seems to have had a good bit of discussion but I didn't see that signoff made explicit. If you can drive that towards a comment here or there and/or link something, that would be good.

I'm particularly concerned about the surface area confusing newer users -- especially given the frequent concerns over async complexity -- I wonder if we should add a sentence/paragraph to the top-level docs in https://doc.rust-lang.org/nightly/std/task/index.html to clarify the current state (i.e., stable surface area should be sufficient for most uses or so).

tvallotton committed:
fix: Apply suggestions from code review

Co-authored-by: Mark Rousskov <mark.simulacrum@gmail.com>

tvallotton committed:
docs: remove recommendations to use LocalWaker in stable API documentation

tvallotton committed:
chore: make method order consistent with waker

tvallotton commented:
> I wonder if we should add a sentence/paragraph to the top-level docs in https://doc.rust-lang.org/nightly/std/task/index.html to clarify the current state (i.e., stable surface area should be sufficient for most uses or so).

I think it may be better to put it on the docs for `LocalWaker`. What do you think?

Thomasdezeeuw commented on library/core/src/task/wake.rs:
```diff
@@ -195,16 +236,33 @@ impl<'a> Context<'a> {
-        &self.waker
+        &self
+            .waker
+            .expect("no waker was set on this context, consider calling `local_waker` instead.")
```
I don't think this is a good idea.

All existing code that uses `Future`s depend on this function returning `Waker`. So any `Context` without a `Waker` will cause this to panic.

tvallotton replied:
Well, the argument is that if a runtime decides not to support `Waker`, they will have to create a waker that panics on `wake()` instead of panicking in the call to `waker()`. Now, several people have expressed the same concern, so I do think it might be best to keep them mandatory for now. It would not be a breaking change to make them optional in the future, so we really don't need to commit to this right now.

Thomasdezeeuw reviewed (changes requested):
I like to see a `LocalWaker` type, but this not the way to do it in my opinion. That's coming from someone who has written a lot of `Future`s, async function and a runtime specialised in thread-locality (i.e. I very much like to see this change).

I don't think making the `Waker` optional in `Context` is wise considering all existing `Future` related code depends on the `Waker` type always being part of the `Context`, it was never not there, a promise we're walking back with this pr.

tvallotton committed:
refactor: make waker mandatory.

This also removes
* impl From<&Context> for ContextBuilder
* Context::try_waker()

The from implementation is removed because now that
wakers are always supported, there are less incentives
to override the current context. Before, the incentive
was to add Waker support to a reactor that didn't have
any.

tvallotton commented:
I made the waker type mandatory on the last commit. Several people have expressed their concern with this aspect of the API. And while I do think that it is better for them to be optional than not, that is not something I consider terribly important.

bors commented:
:umbrella: The latest upstream changes (presumably #120136) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tvallotton committed:
doc: fix some doctests after rebase

tvallotton commented:
@Mark-Simulacrum Do we really need an explicit signoff here? Both teams have discussed it and approved nightly experimentation ([libs-team](https://github.com/rust-lang/libs-team/issues/191#issuecomment-1830225080), [wg-async](https://github.com/rust-lang/libs-team/issues/191#issuecomment-1856383426)). I feel that with no significant changes on the proposal, it is going to be hard to get the teams to meet to discuss the same topic again.

Mark-Simulacrum commented:
Those comments seem sufficient to me (I hadn't seen them in my quick skim) for nightly experimentation.

@rustbot ready

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 180c68bef5f42bd61f517e1a48db4152b6e1989e has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 180c68bef5f42bd61f517e1a48db4152b6e1989e -->
<!-- homu: {"type":"Approved","sha":"180c68bef5f42bd61f517e1a48db4152b6e1989e","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
